stellar_uri.go

```diff
@@ -0,0 +1,336 @@
+
+// ErrInvalidScheme is returned if the URI scheme is not web+stellar.
+var ErrInvalidScheme = errors.New("invalid stellar URI scheme")
+
+// ErrInvalidOperation is returned if the URI operation is not supported.
+var ErrInvalidOperation = errors.New("invalid stellar URI operation")
+
+// ErrBadSignature is returned if the signature fails verification.
+var ErrBadSignature = errors.New("bad signature")
+
+// SignStellarURI signs a stellar+web URI and returns the URI with the signature
+// attached.
+func SignStellarURI(uri string, seed SeedStr) (signedURI, signatureB64 string, err error) {
+	kp, err := keypair.Parse(seed.SecureNoLogString())
+	if err != nil {
+		return "", "", err
+	}
+	payload := payloadFromString(uri)
+	signature, err := kp.Sign(payload)
+	if err != nil {
+		return "", "", err
+	}
+	signatureB64 = base64.StdEncoding.EncodeToString(signature)
+	signatureEsc := url.QueryEscape(signatureB64)
+
+	signedURI = uri + "&signature=" + signatureEsc
+
+	return signedURI, signatureB64, nil
+}
+
+// ValidatedStellarURI contains the origin domain that ValidateStellarURI
+// confirmed
+type ValidatedStellarURI struct {
+	URI          string
+	Operation    string
+	OriginDomain string
+	Message      string
+	CallbackURL  string
+	XDR          string
+	TxEnv        *xdr.TransactionEnvelope
+	Recipient    string
+	Amount       string
+	AssetCode    string
+	AssetIssuer  string
+	Memo         string
+	MemoType     string
+}
+
+// ValidateStellarURI will check the validity of a web+stellar SEP7 URI.
+//
+// It will check that the parameters are valid and that the payload is
+// signed with the appropriate key.
+func ValidateStellarURI(uri string, getter HTTPGetter) (*ValidatedStellarURI, error) {
+	uv, err := newUnvalidatedURI(uri)
+	if err != nil {
+		return nil, err
+	}
+	return uv.Validate(getter)
+}
+
+type unvalidatedURI struct {
+	raw          string
+	values       url.Values
+	Operation    string
+	OriginDomain string
+	Signature    string
+}
+
+func newUnvalidatedURI(uri string) (*unvalidatedURI, error) {
+	res := &unvalidatedURI{raw: uri}
+
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme != "web+stellar" {
+		return nil, ErrInvalidScheme
+	}
+
+	res.Operation = u.Opaque
+
+	res.values = u.Query()
+
+	res.OriginDomain = res.value("origin_domain")
+	res.Signature = res.value("signature")
+
+	return res, nil
+
+}
+
+func (u *unvalidatedURI) Validate(getter HTTPGetter) (*ValidatedStellarURI, error) {
+	// origin_domain and signature are optional in the spec, but
+	// it seems like a really bad idea to allow any of these
+	// requests without them, so we are going to make them required.
+	if u.OriginDomain == "" {
+		return nil, ErrMissingParameter{Key: "origin_domain"}
+	}
+	if u.Signature == "" {
+		return nil, ErrMissingParameter{Key: "signature"}
+	}
+
+	// make sure there's no port or scheme
+	if strings.IndexByte(u.OriginDomain, ':') != -1 {
+		return nil, ErrInvalidParameter{Key: "origin_domain"}
+	}
```

Comment:
Should block IDN homograph attacks like "exаmple.com" (that's no "a"). If there's nothing else doing that right now, a blunt solution would be to allow only a subset of ascii characters in OriginDomain here. Or convert it to the `xn-`-encoding

Reply:
spec says should support not just ascii.

> Suggested Precaution: Wallets should be very careful about the fonts they use when displaying the origin_domain, memo and msg. Some fonts do not clearly distinguish between certain characters so it is easy to confuse these characters. Wallets should use a font that allows to clearly distinguish between characters so that users can clearly identify malicious domains that are pretending to be trusted domains. Consider testing your wallet's UI with characters ranging across Unicode (particularly Cyrillic) and not just ASCII.

I'm not sure a font is the best solution to this, but...

Reply:
I agree, don't think a font can hope to solve all the `а != a` problems. Supporting only ascii seems better than presenting exаmple.com as a verified domain.

Reply:
I added `isDomainName` to check the origin_domain.  I'm assuming the wording here in the spec to support cyrillic is for the memo and msg.